LRU cache prim

tfeher opened the pull request:
SW managed cache in device memory, for ML algos where we can trade memory access for computation.

teju85 commented on cpp/src_prims/cache/cache.h:
```diff
@@ -0,0 +1,314 @@
+* the cache buffer, where a unit of data that we are storing is math_t[n_vec].
+* For example in SVM we store full columns of the kernel matrix at each cache
+* entry.
+*
```
Minor knitpick: adding a usage example here in the documentation would be very useful, just like done in this header file: https://github.com/rapidsai/cuml/blob/branch-0.9/cpp/src_prims/common/device_buffer.hpp

tfeher replied:
That is a very good point, I have added an example.

tfeher commented on cpp/src_prims/cache/cache.h:
```diff
@@ -0,0 +1,314 @@
+      n_cache_sets = 0;
+      cache_size = 0;
+    }
+    //std::cout << "Creating cache with size " << cache_size << " MiB, to store " <<
```
Thanks @cjnolet for the comments! Ideally this message would be logged with INFO level. I could remove it, but I would keep it as a reminder that I wanted to log something here. This issue comes up several times in PR #912 too. What is the best practice until we decide about logging #100?

cjnolet replied:
The convention in sklearn is to pass a boolean `verbose` argument down to algorithms to enable logging like this. This could further be passed down to the prims.

Will this strategy work for you until we have better logging in place?

tfeher replied:
That should work, I'll implement it tomorrow morning.

tfeher commented on cpp/src_prims/cache/cache.h:
```diff
@@ -0,0 +1,314 @@
+class Cache {
+ public:
+  /**
+   * Construct a Cache object
```
Yes, it seems autobrief is off, so you are right. I am adding the brief tags.

tfeher commented on cpp/src_prims/cache/cache_util.h:
```diff
@@ -0,0 +1,342 @@
+ * @param [out] is_cached  whether the element is cached size[n]
+ * @param [in] time iteration counter (used for time stamping)
+ */
+inline __global__ void get_cache_idx(int *idx, int n, int *vec_idx,
```
A minor issue: the `inline` specifier is used here to indicate that there could be multiple definitions of this non-template function, when the header is included in multiple translation units (otherwise we would get multiple definition error during linking). AFAIK this is a valid use case for the `inline` specifier.

The compiler emits the following `warning: inline qualifier ignored for "__global__" function`. This is understood and can be ignored, since we do not intend the global function to be inlined. 

If the warning message is disturbing, then we have the following options: 
 - add the following compiler option `-Xcudafe="--diag_suppress=2943" `. (In general one can display the error/warning message numbers with the `-Xcudafe="--display_error_number"` flag, and use the diag_suppress option to disable it)
 - use `static` keyword instead of `inline`
 - put the function in an unnamed namespace

Let me know which option would you prefer.

teju85 replied:
I'd prefer unnamed namespace.

cjnolet commented:
@tfeher Have all the review feedback items been addressed in this PR? Is this ready for merge?

tfeher commented:
@cjnolet Yes, I have addressed all the feedback. You could have a look in the example usage (cache.h, line 54), if that is clear enough. From my side it is ready for merge.

cjnolet commented:
@tfeher, I had addressed a few additional items in your SVM that i had originally missed in this PR. 

Are you okay with closing this PR and having the cache prim merged w/ the SVM PR?

tfeher commented:
@cjnolet, yes we can do that. And sorry, I have realized that I was not using the review features of GitHub properly for this PR, I think that caused some confusion. I will merge the changes from here into SVM, and we can continue there.
